behavior/src/behavior.ts

```diff
@@ -0,0 +1,239 @@
+type A<T> = ReadonlyArray<T>
+
+// Priorities for overriding the ordering of extensions and behaviors.
+export enum Priority { fallback = -1, base = 0, extend = 1, override = 2 }
+
+const noPriority = -2 as Priority
+
+const none = [] as any
+
+// A behavior is a type of value that can be associated with an
+// extendable object. The `Value` type parameter determines the type
+// of value this behavior holds, and the `Target` type parameter
+// describes the kind of value that this extends.
+export interface Behavior<Value, Target> {
+  (value: Value, priority?: Priority): Extender<Target>
+}
+
+// Define a new behavior. When `unique` is true, this behavior is
+// intended to exist at most once per target value.
+export function defineBehavior<Value, Target>(unique: boolean = false): Behavior<Value, Target> {
+  let behavior = function(value: Value, priority: Priority = noPriority): Extender<Target> {
+    return new Extender(null, behavior, value, priority)
+  }
+  ;(behavior as any).unique = unique
```

Comment:
Wouldn't it be nicer to put `unique` in the interface?

Reply:
I figured it'd not be something that user code would ever really need to query, and everything in the interface will be public, so this is my way of making it private. (Possibly using `@internal` would also work here.)

Reply:
Since it's used in `state/` it doesn't feel private or even `@internal`.

Reply:
How did you handle such cases in ProseMirror?

Reply:
Oh, that just needs some refactoring (I was planning to add a `getSingle` method to `BehaviorStore` but forgot).

In ProseMirror, only the things that get doc comments with `::` end up in the docs, so it's easier to hide module-private stuff.

Reply:
Oh okay, if you move that usage to `behavior` I'm fine with it :)